node/src/components/block_synchronizer/tests.rs

```diff
@@ -95,35 +180,77 @@ fn check_sync_global_state_event(event: MockReactorEvent, block: &Block) {
     );
 }
 
+// Calls need_next for the block_synchronizer and processes the effects resulted returning a list of
+// the new events that were generated
+async fn need_next(
+    rng: &mut TestRng,
+    reactor: &MockReactor,
+    block_synchronizer: &mut BlockSynchronizer,
+    num_expected_events: usize,
+) -> Vec<MockReactorEvent> {
+    let effects = block_synchronizer.need_next(reactor.effect_builder(), rng);
+    assert_eq!(effects.len(), num_expected_events);
+    reactor.process_effects(effects).await
+}
+
+fn register_multiple_signatures<'a, I: IntoIterator<Item = &'a Arc<SecretKey>>>(
+    builder: &mut BlockBuilder,
+    block: &Block,
+    validator_keys_iter: I,
+) {
+    for secret_key in validator_keys_iter {
+        // Register a finality signature
+        let signature = FinalitySignature::create(
+            *block.hash(),
+            block.header().era_id(),
+            secret_key.as_ref(),
+            PublicKey::from(secret_key.as_ref()),
+        );
+        assert!(signature.is_verified().is_ok());
+        assert!(builder.register_finality_signature(signature, None).is_ok());
+    }
+}
+
+impl BlockSynchronizer {
+    // Create an initialized block synchronizer with default config and MAX_SIMULTANEOUS_PEERS peers
+    fn new_initialized(rng: &mut TestRng, validator_matrix: ValidatorMatrix) -> BlockSynchronizer {
+        let mut block_synchronizer = BlockSynchronizer::new(
+            Config::default(),
+            Arc::new(Chainspec::random(rng)),
+            MAX_SIMULTANEOUS_PEERS as u32,
+            validator_matrix,
+            &prometheus::Registry::new(),
+        )
+        .expect("Failed to create BlockSynchronizer");
+
+        <BlockSynchronizer as InitializedComponent<MainEvent>>::set_state(
+            &mut block_synchronizer,
+            ComponentState::Initialized,
+        );
+
+        block_synchronizer
+    }
+
+    fn forward_builder(&self) -> &BlockBuilder {
+        self.forward.as_ref().expect("Forward builder missing")
+    }
+}
+
 #[tokio::test]
 async fn global_state_sync_wont_stall_with_bad_peers() {
     let mut rng = TestRng::new();
     let mock_reactor = MockReactor::new();
-
-    // Set up a random block that we will use to test synchronization
-    let block = Block::random(&mut rng);
-
-    // Set up a validator matrix for the era in which our test block was created
-    let mut validator_matrix = ValidatorMatrix::new_with_validator(ALICE_SECRET_KEY.clone());
-    validator_matrix.register_validator_weights(
-        block.header().era_id(),
-        iter::once((ALICE_PUBLIC_KEY.clone(), 100.into())).collect(),
+    let test_env = TestEnv::random(&mut rng).with_block(
+        TestBlockBuilder::new()
+            .era(1)
+            .deploys([Deploy::random(&mut rng)].iter())
```

Comment:
Would be useful to add a `random_deploys(count)` method on the builder (I recently used something similar too), but it's not really necessary to do it in this PR.

Reply:
Added here: https://github.com/casper-network/casper-node/pull/3816/commits/6404ee17f8ffc9e42e42ddf19f905b214791ef02